core/frontend/src/tools/ElementSetTool.ts

```diff
@@ -150,44 +60,41 @@ export class ElementAgenda {
   public get count() { return this.length; }
   public get length() { return this.elements.length; }
 
-  /** Calls ClearHilite and empties this ElementAgenda. */
-  public clear() { this.clearHilite(); this.elements.length = 0; this.groupMarks.length = 0; }
+  /** Empties this ElementAgenda and clears hilite state when manageHiliteState is true. */
+  public clear() { this.setEntriesHiliteState(false); this.elements.length = 0; this.groupMarks.length = 0; }
 
-  /** clear hilite on any currently hilited entries */
-  private clearHilite() {
-    if (HilitedState.No === this.hilitedState)
+  private setEntriesHiliteState(onOff: boolean, groupStart = 0, groupEnd = 0) {
+    if (!this.manageHiliteState)
       return;
 
-    this.setEntriesHiliteState(false); // make sure all entries have their hilite flag off
-    this.hilitedState = HilitedState.No;
-  }
+    const ss = this.iModel.selectionSet.isActive ? this.iModel.selectionSet.elements : undefined;
+    if (undefined === ss && 0 === groupEnd) {
+      this.iModel.hilited.setHilite(this.elements, onOff);
+      return;
+    }
 
-  private setEntriesHiliteState(onOff: boolean, groupStart = 0, groupEnd = 0) {
-    const group = (0 === groupEnd) ? this.elements : this.elements.filter((_id, index) => index >= groupStart && index < groupEnd);
+    const shouldChangeHilite = (id: string, index: number) => {
+      if (undefined !== ss && ss.has(id))
+        return false; // Don't turn hilite on/off for elements in current selection set...
+      return (0 === groupEnd || (index >= groupStart && index < groupEnd));
+    };
+
+    const group = this.elements.filter((id, index) => shouldChangeHilite(id, index));
     this.iModel.hilited.setHilite(group, onOff);
   }
 
-  /** Calls ClearHilite and removes the last group of elements added to this ElementAgenda. */
+  /** Removes the last group of elements added to this ElementAgenda. */
   public popGroup() {
     if (this.groupMarks.length <= 1) {
       this.clear();
       return;
     }
     const group = this.groupMarks.pop()!;
-    if (HilitedState.No !== this.hilitedState)
-      this.setEntriesHiliteState(false, group.start, this.length); // make sure removed entries aren't left hilited...
+    this.setEntriesHiliteState(false, group.start, this.length); // make sure removed entries aren't left hilited...
     this.elements.splice(group.start);
   }
 
-  /** Mark all entries in this agenda as being hilited. */
-  public hilite() {
-    if (HilitedState.Yes === this.hilitedState)
-      return;
-
-    this.setEntriesHiliteState(this.hiliteOnAdd); // make sure all entries have their hilite flag on.
-    this.hilitedState = HilitedState.Yes;
-  }
-
+  /** Return true if elementId is already in this ElementAgenda. */
```

Comment:
`add` calls this in a loop to make sure a given element only appears once in the array. Doesn't the agenda really want to be a `Set` not an array?

Reply:
No, the array is important for order, ex. tool wants 1st to be a solid, 2nd to be a surface, etc.

Reply:
`Set` preserves order.

Reply:
Only being able to iterate in insert order seemed inconvenient, and I'd have to rework GroupMark?